src/ir_passes/compile.rs

```diff
@@ -0,0 +1,754 @@
+            ir::Expr::Bin { .. } | ir::Expr::Fn { .. } => {
+                panic!("Port width must be a parameter or constant.")
+            }
+        }
+    }
+
+    /// Compiles an [ir::ExprIdx] into a [u64].
+    /// Expects the [ir::ExprIdx] to be a single constant value
+    fn u64(comp: &ir::Component, expr: ir::ExprIdx) -> u64 {
+        match comp.get(expr) {
+            ir::Expr::Concrete(val) => *val,
+            ir::Expr::Param(..)
+            | ir::Expr::Bin { .. }
+            | ir::Expr::Fn { .. } => {
+                panic!("Port width must be a constant.")
+            }
+        }
+    }
+
+    fn ports<CW, WFU, WT>(
+        ctx: &ir::Context,
+        comp: &ir::Component,
+        wfu: WFU, // Function that returns a CW type from a u64
+        width_transform: WT,
+        is_comp: bool,
+    ) -> Vec<calyx::PortDef<CW>>
+    where
+        WFU: Fn(u64) -> CW,
+        WT: Fn(&ir::Component, ir::ExprIdx) -> CW,
+    {
+        let mut ports: Vec<_> = comp
+            .ports()
+            .idx_iter()
+            .filter_map(|port| {
+                Compile::port_def(ctx, comp, port, &width_transform)
+            })
+            .chain(comp.unannotated_ports().into_iter().map(|(name, width)| {
+                calyx::PortDef {
+                    name: name.as_ref().into(),
+                    width: wfu(width),
+                    direction: calyx::Direction::Input,
+                    attributes: calyx::Attributes::default(),
+                }
+            }))
+            .chain(
+                comp.events()
+                    .idx_iter()
+                    .filter_map(|idx| Compile::interface_port_name(comp, idx))
+                    .into_iter()
+                    .map(|name| calyx::PortDef {
+                        name: name.into(),
+                        width: wfu(1),
+                        direction: calyx::Direction::Input,
+                        attributes: vec![(
+                            calyx::Attribute::Unknown("fil_event".into()),
+                            1,
+                        )]
+                        .try_into()
+                        .unwrap(),
+                    }),
+            )
+            .collect();
+
+        let mut interface_ports =
+            INTERFACE_PORTS.iter().collect::<HashSet<_>>();
+
+        // add interface port attributes if necessary
+        for pd in &mut ports {
+            if let Some(pair) = INTERFACE_PORTS
+                .iter()
+                .find(|(_, (n, _, _))| *n == pd.name.as_ref())
+            {
+                assert!(
+                    pair.1 .2 == pd.direction,
+                    "Expected {} to be an {:?} port, got {:?} port.",
+                    pair.1 .0,
+                    pair.1 .2,
+                    pd.direction
+                );
+                // TODO: Assert width equality
+                interface_ports.remove(pair);
```

Comment:
This `TODO` item needs more details if we're going to merge the changes with TODO. Alternatively, address the TODO